File: CS-320/Java.pre-modifications/src/test/java/snhu/ContactServiceTest.java
package snhu;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ContactServiceTest {

    /**
     * Tests for the addContact method in the ContactService class.
     * The addContact method is responsible for adding a new contact into the storage.
     */
    @Test
    void testAddContact() {
        ContactService contactService = new ContactService();
        Contact contact1 = new Contact("first", "last", "1234567890", "address");
        // We can add a valid contact
        assertDoesNotThrow(() -> {
            contactService.addContact(contact1);
        });
        // Throw exception when contact ID already exists
        IllegalArgumentException error1 = assertThrows(IllegalArgumentException.class, () -> {
            contactService.addContact(contact1);
        });
        assertEquals("Contact with ID " + contact1.getId() + " already exists", error1.getMessage());
        // Throw exception when the contact is not valid
        IllegalArgumentException error2 = assertThrows(IllegalArgumentException.class, () -> {
            Contact contact = new Contact(null, null, null, null);
            contactService.addContact(contact);
        });
        assertEquals("First name is required", error2.getMessage());
    }

    /**
     * Tests for the deleteContact method in the ContactService class.
     * The deleteContact method is responsible for removing an existing contact from the storage.
     */
    @Test
    void testDeleteContact() {
        ContactService contactService = new ContactService();
        Contact contact = new Contact("first", "last", "1234567890", "address");
        contactService.addContact(contact);
        // We can remove an existing contact
        assertDoesNotThrow(() -> {
            contactService.deleteContact(contact.getId());
        });
        // Throw exception when the contact does not exist
        IllegalArgumentException error = assertThrows(IllegalArgumentException.class, () -> {
            contactService.deleteContact(contact.getId());
        });
        assertEquals("Contact with ID " + contact.getId() + " does not exist", error.getMessage());
    }

    /**
     * Tests for the updateContact method in the ContactService class.
     * The updateContact method is responsible for updating an existing contact in the storage.
     */
    @Test
    void testUpdateContact() {
        ContactService contactService = new ContactService();
        Contact contact = new Contact("first", "last", "1234567890", "address");
        contactService.addContact(contact);
        // We can update an existing contact
        assertDoesNotThrow(() -> {
            contactService.updateContact(contact.getId(), "new first", "new last", "1234567891", "new address");
        });
        // Throw exception when the contact does not exist
        IllegalArgumentException error1 = assertThrows(IllegalArgumentException.class, () -> {
            contactService.deleteContact("0");
        });
        assertEquals("Contact with ID 0 does not exist", error1.getMessage());
        // Throw exception when the contact update makes it not valid
        IllegalArgumentException error2 = assertThrows(IllegalArgumentException.class, () -> {
            contactService.updateContact(contact.getId(), null, null, null, null);
        });
        assertEquals("First name is required", error2.getMessage());
    }

}

File: CS-320/Go.post-modifications/contact/contact_service_test.go
package contact

import (
	"strings"
	"testing"
)

// TestAddContact tests the AddContact method of the ContactService
func TestAddContact(t *testing.T) {
	contactService := NewContactService()

	// Test adding a valid contact
	contact := NewContactFull("1", "first", "last", "1234567890", "address")
	err := contactService.AddContact(contact)
	if err != nil {
		t.Errorf("Expected no error when adding valid contact, got: %v", err)
	}

	// Test adding a contact with duplicate ID
	duplicateContact := NewContactFull("1", "first", "last", "1234567890", "address")
	err = contactService.AddContact(duplicateContact)
	if err == nil {
		t.Error("Expected error when adding contact with duplicate ID, got nil")
	}
	expectedErrMsg := "contact with ID 1 already exists"
	if err != nil && !strings.Contains(err.Error(), expectedErrMsg) {
		t.Errorf("Expected error message '%s', got: %v", expectedErrMsg, err)
	}

	// Test adding an invalid contact
	invalidContact := NewContact("")
	err = contactService.AddContact(invalidContact)
	if err == nil {
		t.Error("Expected error when adding invalid contact, got nil")
	}
	expectedErrMsg = "ID is required"
	if err != nil && !strings.Contains(err.Error(), expectedErrMsg) {
		t.Errorf("Expected error message '%s', got: %v", expectedErrMsg, err)
	}
}

// TestDeleteContact tests the DeleteContact method of the ContactService
func TestDeleteContact(t *testing.T) {
	contactService := NewContactService()
	contact := NewContactFull("1", "first", "last", "1234567890", "address")

	// Setup: Add a contact
	err := contactService.AddContact(contact)
	if err != nil {
		t.Fatalf("Failed to set up test: %v", err)
	}

	// Test deleting an existing contact
	err = contactService.DeleteContact(contact.ID)
	if err != nil {
		t.Errorf("Expected no error when deleting existing contact, got: %v", err)
	}

	// Test deleting a non-existent contact
	err = contactService.DeleteContact(contact.ID)
	if err == nil {
		t.Error("Expected error when deleting non-existent contact, got nil")
	}
	expectedErrMsg := "Contact with ID 1 does not exist"
	if err != nil && !strings.Contains(err.Error(), expectedErrMsg) {
		t.Errorf("Expected error message '%s', got: %v", expectedErrMsg, err)
	}
}

// TestUpdateContact tests the UpdateContact method of the ContactService
func TestUpdateContact(t *testing.T) {
	contactService := NewContactService()
	contact := NewContactFull("1", "first", "last", "1234567890", "address")

	// Setup: Add a contact
	err := contactService.AddContact(contact)
	if err != nil {
		t.Fatalf("Failed to set up test: %v", err)
	}

	// Test updating an existing contact
	err = contactService.UpdateContact(contact.ID, "new first", "new last", "1234567891", "new address")
	if err != nil {
		t.Errorf("Expected no error when updating existing contact, got: %v", err)
	}

	// Test updating a non-existent contact
	err = contactService.DeleteContact("2")
	if err == nil {
		t.Error("Expected error when deleting non-existent contact, got nil")
	}
	expectedErrMsg := "Contact with ID 2 does not exist"
	if err != nil && !strings.Contains(err.Error(), expectedErrMsg) {
		t.Errorf("Expected error message '%s', got: %v", expectedErrMsg, err)
	}

	// Test updating with invalid data
	err = contactService.UpdateContact(contact.ID, "", "", "", "")
	if err == nil {
		t.Error("Expected error when updating with invalid data, got nil")
	}
	expectedErrMsg = "First name is required"
	if err != nil && !strings.Contains(err.Error(), expectedErrMsg) {
		t.Errorf("Expected error message '%s', got: %v", expectedErrMsg, err)
	}
}
